src/utils/recipient.js

```diff
@@ -26,6 +26,18 @@ export const recipientsOfPrivateMessage = (
   return recipients;
 };
 
+/**
+ * A set of users identifying a PM conversation, as per pmKeyRecipientsFromMessage.
+ *
+ * This is an "opaque type alias" for an array of plain old data.  Code
+ * outside this module can freely inspect and use the data from a value of
+ * this type, but only code within this module can create one.  See Flow docs:
+ *   https://flow.org/en/docs/types/opaque-types/
+ *
```

Comment:
> I think it's the `: $ReadOnlyArray<PmRecipientUser>` bit that accomplishes that, right? (If I remove it at this commit, the `.map` call in `pmNarrowFromRecipients` becomes suspect.)

Yep! That's the "subtyping constraint", in that Flow doc's terminology.


> "freely inspect and use" means freely inspect and use _as a subtype of `$ReadOnlyArray<PmRecipientUser>`_; do I understand that correctly?

That's right. But because that's also the underlying type (the type that `PmKeyRecipients` is secretly an alias for), that's the same thing as saying it can freely inspect and use all the data inside it.

Another thing we could have done -- which hmm, really that doc should have an example for -- is give a bound / a "subtyping constraint" which is not quite as specific as the underlying type. Like this:
```js
export opaque type PmKeyRecipients: $ReadOnlyArray<{ +email: string, ... }> =
  $ReadOnlyArray<PmRecipientUser>;
```
I.e., we'd say "this is an array of some kind of object, where you can read `.email` and get a string; and there might be more but that's all you get to know about."

Then that `.map` call:
```js
export const pmNarrowFromRecipients = (recipients: PmKeyRecipients): Narrow =>
  pmNarrowFromEmails(recipients.map(r => r.email));
```
would still be OK, because it only needs the `email` property. But we couldn't e.g. look at the `full_name` property, even though it's there on `PmKeyRecipients` and therefore should definitely be there on the actual objects at runtime.

Reply:
Expanded this jsdoc for my upcoming revision.